crates/red_knot_python_semantic/src/semantic_index/re_exports.rs

```diff
@@ -0,0 +1,299 @@
+//! A visitor and query to find all global-scope symbols that are exported from a module
+//! when a wildcard import is used.
+//!
+//! For example, if a module `foo` contains `from bar import *`, which symbols from the global
+//! scope of `bar` are imported into the global namespace of `foo`?
+
+use ruff_db::{files::File, parsed::parsed_module};
+use ruff_python_ast::{
+    self as ast,
+    name::Name,
+    visitor::{walk_expr, walk_pattern, walk_stmt, Visitor},
+};
+use rustc_hash::FxHashSet;
+
+use crate::{module_name::module_name_from_import_statement, resolve_module, Db};
+
+#[salsa::tracked(return_ref)]
+pub(super) fn find_exports(db: &dyn Db, file: File) -> FxHashSet<Name> {
+    let module = parsed_module(db.upcast(), file);
+
+    let mut finder = ExportFinder {
+        db,
+        file,
+        exports: FxHashSet::default(),
+    };
+
+    finder.visit_body(module.suite());
+    finder.exports
+}
+
+struct ExportFinder<'db> {
+    db: &'db dyn Db,
+    file: File,
+    exports: FxHashSet<Name>,
+}
+
+impl<'db> ExportFinder<'db> {
+    fn possibly_add_export(&mut self, name: &Name) {
+        if name.starts_with('_') {
+            return;
+        }
+        if !self.exports.contains(name) {
```

Comment:
The `contains` + `insert` check has the downside that `name` gets hashed twice. I assume you do it this way to avoid cloning `name` when the name already exists (although that's probably unlikely?). https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.get_or_insert_with would be nice but it's nightly only.

I'm leaning towards always cloning. Or you can use a `FxHashMap<Name, ()>` with the entry API?

Reply:
Yeah, I thought you might have things to say about this ;)

I also wasn't sure if this was worth it. I switched it to just do the clone every time.

I tried the `FxHashMap` idea but as far as I can tell the `entry` API also requires the value passed in to be owned rather than be a reference? So I'm not sure that helps us here